docs/faq/env_var.md

```diff
@@ -280,6 +280,11 @@ When USE_PROFILER is enabled in Makefile or CMake, the following environments ca
   - Values: Int ```(default=4)```
   - This variable controls how many CuDNN dropout state resources to create for each GPU context for use in operator.
 
+* MXNET_ENFORCE_SAFE_ACCUMULATION
+  - Values: Values: 0(false) or 1(true) ```(default=0)```
+  - If this variable is set, the accumulation will enter the safe mode, meaning accumulation is done in a data type of higher precision than
+    the input data type, leading to more accurate results with a possible performance loss and backward compatibility loss.
```

Comment:
Can we mention the list of operator supporting this env_var?

Reply:
Also can you mention what casting happens ie., float16->float32, etc.,

Reply:
Also I am concerned about float32-> float64 casting as described in https://github.com/apache/incubator-mxnet/issues/14722.
In our offline discussion @haojin2 and @eric-haibin-lin seem to allude that somehow now with float64 casting for accumulation it is exposing a higher loss which is the correct value and the earlier model which yields lower loss is actually wrong. I am not sure I understand this despite our long conversation (sorry!). 
Also I found another [old PR](https://github.com/dmlc/mshadow/pull/236) which casted the Accumulation variable to higher type, however [comment] (https://github.com/dmlc/mshadow/pull/236#discussion_r113569026) seem to suggest float32->float32 is sufficient(I checked with @cjolivier01 for the reasoning he does not remember unfortunately).
My suggestion is to retain float32->float32 unless you have a convincing argument.

Reply:
@nswamy I don't see a concrete reason why you want fp32 accumulation type for fp32 inputs if user set MXNET_ENFORCE_SAFE_ACCUMULATION=1. The only reason I see from the mshadow PR is to keep backward compatibility. 
If you use fp32 accumulation type for fp32 inputs, it is NOT safe accumulation. 

One simple test case where fp32 accumulation is not sufficient for safe accumulation:
```
# fp32 accumulation
>>> import numpy as np
>>> val1 = np.float32(3e38)
>>> val2 = np.float32(-3e38)
>>> sum = np.float32(3e38)
>>> sum + val1 + val2
inf
>>> sum64 = np.float64(3e38)
>>> sum64 + val1 + val2
3e+38
>>> np.float32(sum64 + val1 + val2)
3e+38
```

Reply:
@nswamy Although I agree that using fp32 for fp32 accumulation may be sufficient under many cases, but:
1. As @eric-haibin-lin said you cannot claim that such accumulation is safe.
2. Usually a user is not 100% sure that his/her data will not cause overflow during any accumulation, so we want to provide an option if users want to be safe even he/she is using fp32 data type.

Reply:
https://github.com/pytorch/pytorch/blob/cca247635c6edb323176eeac7a18d3e9ab71c558/torch/lib/THD/base/THDGenerateAllTypes.h#L70-L71

Reply:
@haojin2 @eric-haibin-lin do you have an explanation for why @nswamy 's example the loss would go up with fp32 inputs and accumulating in fp64?

Reply:
@anirudh2290 there's no guarantee that being more precise gives you a model with better accuracy / lower perplexity.

Reply:
I understand that there is no guarantee of getting better accuracy, but can it be worse and what are reasons for the same ? This is what seems to be happening for @nswamy 's example.

Reply:
yeah it's possible that the model gets worse. The accuracy could go either way

Reply:
In this case we need to change the env variable documentation: "leading to more accurate results with a possible performance loss and backward compatibility loss."

Reply:
What does backward compatibility loss mean?